php/Blockonomics.php

```diff
@@ -131,59 +131,78 @@ public function check_get_callbacks_response_code($response){
+        $matching_callback = false;
+        $different_protocols = false;
+        $different_protocol_xpub = '';
+        $available_xpub = '';
+        $error_str = '';
+
+        //go through all xpubs on the server and examine their callback url
+        foreach($response_body as $one_response){
+            $server_callback_url = isset($one_response->callback) ? $one_response->callback : '';
+            $xpub = isset($one_response->address) ? $one_response->address : '';
+            //if the xpub doesn't have a callback => store available xpub
+            if(!$server_callback_url){
+                $available_xpub = $xpub;
+            //Exact match => return
+            }else if($server_callback_url == $wordpress_callback_url){
+                return $error_str;
+            //All but secret match => update_callback and return
+            }else if(strpos($server_callback_url, $api_url) === 0){
+                $this->update_callback($wordpress_callback_url, $crypto, $xpub);
+                return $error_str;
             }
-            elseif($response_callback_without_schema != $callback_url_without_schema)
-            {
-                $base_url = get_bloginfo('wpurl');
-                $base_url = preg_replace('/https?:\/\//', '', $base_url);
-                // Check if only secret differs
-                if(strpos($response_callback, $base_url) !== false)
-                {
-                    //Looks like the user regenrated callback by mistake
-                    //Just force Update_callback on server
-                    $this->update_callback($callback_url, $crypto, $response_address);
-                }
-                else
-                {
-                    $error_str = __("You have an existing callback URL", 'blockonomics-bitcoin-payments');
-                }
-                
+            //matching url but the protocols are different => store values and only update as last resort
+            else if(strpos($server_callback_url, $base_url)){
+                $different_protocols = true;
+                $different_protocol_xpub = $xpub;
             }
         }
-        else 
-        {
-            $error_str = __("You have an existing callback URL", 'blockonomics-bitcoin-payments');
-            // Check if callback url is set
-            foreach ($response_body as $res_obj)
-             if(preg_replace('/https?:\/\//', '', $res_obj->callback) == $callback_url_without_schema)
-                $error_str = "";
-        }  
+
+        //Scenario 1 (Exact match) and 2 (All but secret matches) failed => use available xpub if we have one  
+        if($available_xpub){
+            $this->update_callback($wordpress_callback_url, $crypto, $available_xpub);
+            $matching_callback = true;
+        //Scenario 1 (Exact match) and 2 (All but secret matches) failed, also no available_xpub => use callback with different protocol
+        } elseif ($different_protocols){
```

Comment:
If there is an _available_xpub_ and an xpub with a callback URL that has _different_protocols_, the callback with a different protocol should be updated and used. In this case, the available_xpub is used, which can lead to `xpub1 - http://store.com/?secret=123` and `xpub2 - https://store.com/?secret=123` instead of updating xpub1.

In any scenario, if a partly matching domain is found it should be updated and test setup return success. To keep the code simple we only need 3 checks
```
  // No callback
  if(!$server_callback_url){
    $available_xpub = $xpub;
  }
  // Exact match
  }else if($server_callback_url == $wordpress_callback_url){
    return '';
  }
  // Partial Match - Only secret or protocol differ etc
  else if(strpos($server_callback_url, $api_url) === 0 || strpos($server_callback_url, $base_url)){
    $this->update_callback($wordpress_callback_url, $crypto, $xpub);
    return '';
  }
}

// Use the available xpub
if($available_xpub){
  $this->update_callback($wordpress_callback_url, $crypto, $available_xpub);
  return '';
}
// No match and no empty callback
$error_str = __("Multiple callback error: Please add a new store with valid xpub", 'blockonomics-bitcoin-payments');
return $error_str;
```
This removes `$different_protocols` and `$different_protocol_xpub` as we update and return immediately. We can also remove `$matching_callback`  and `$error_str = '';` as we return an empty string instead and only use `$error_str` for the error.

Reply:
I agree, if we don't check the protocols the algorithm becomes more streamlined.

Reply:
We can still check if only the protocol differs to constitute a Partial Match. Different secret or different protocol can be handled the same. ie update and return success:
```
// Partial Match - Only secret or protocol differ etc
else if(strpos($server_callback_url, $api_url) === 0 || strpos($server_callback_url, $base_url)){
  $this->update_callback($wordpress_callback_url, $crypto, $xpub);
  return '';
}
```

Reply:
Sure, just thought that it was easier to keep it to `strpos($server_callback_url, $base_url)`, since it will catch both different secrets and different protocols. The end result should be the same?

However, I've re-introduced the `strpos($server_callback_url, $api_url) === 0` check again.